File: src/whp/vcpu_state.cpp
#include "vcpu_state.h"

#include "common.h"

#include <algorithm>
#include <cstdio>
#include <cstring>
#include <exception>
#include <span>
#include <vector>

namespace tinyvmm::whp::snapshot {

// Architectural register set. Order is part of the snapshot ABI in
// VCPU_REGS sections, but each WHV_REGISTER_NAME is tagged on the wire
// so we can safely re-order at read time; the on-write order is just
// "what the writer happened to choose". Keep this list stable so newly
// captured snapshots remain reasonable to diff against older ones.
const WHV_REGISTER_NAME kArchRegNames[] = {
    // GPRs
    WHvX64RegisterRax, WHvX64RegisterRcx, WHvX64RegisterRdx, WHvX64RegisterRbx,
    WHvX64RegisterRsp, WHvX64RegisterRbp, WHvX64RegisterRsi, WHvX64RegisterRdi,
    WHvX64RegisterR8,  WHvX64RegisterR9,  WHvX64RegisterR10, WHvX64RegisterR11,
    WHvX64RegisterR12, WHvX64RegisterR13, WHvX64RegisterR14, WHvX64RegisterR15,
    WHvX64RegisterRip, WHvX64RegisterRflags,
    // Segments
    WHvX64RegisterEs, WHvX64RegisterCs, WHvX64RegisterSs, WHvX64RegisterDs,
    WHvX64RegisterFs, WHvX64RegisterGs, WHvX64RegisterLdtr, WHvX64RegisterTr,
    // Tables
    WHvX64RegisterIdtr, WHvX64RegisterGdtr,
    // Control
    WHvX64RegisterCr0, WHvX64RegisterCr2, WHvX64RegisterCr3, WHvX64RegisterCr4,
    WHvX64RegisterCr8, WHvX64RegisterXCr0,
    // Debug
    WHvX64RegisterDr0, WHvX64RegisterDr1, WHvX64RegisterDr2,
    WHvX64RegisterDr3, WHvX64RegisterDr6, WHvX64RegisterDr7,
    // MSR-as-register (excluding TSC family — those live in kTimingRegNames
    // so they can be applied LAST during restore)
    WHvX64RegisterEfer, WHvX64RegisterKernelGsBase, WHvX64RegisterApicBase,
    WHvX64RegisterPat,
    WHvX64RegisterSysenterCs, WHvX64RegisterSysenterEsp, WHvX64RegisterSysenterEip,
    WHvX64RegisterStar, WHvX64RegisterLstar, WHvX64RegisterCstar,
    WHvX64RegisterSfmask,
};

const WHV_REGISTER_NAME kTimingRegNames[] = {
    WHvX64RegisterTsc,
    WHvX64RegisterTscAux,
};

const WHV_REGISTER_NAME kIntrCtlRegNames[] = {
    WHvRegisterPendingInterruption,
    WHvRegisterInterruptState,
    WHvRegisterDeliverabilityNotifications,
    WHvRegisterInternalActivityState,
};

// M33.7: Supervisor MSRs. Order is part of the snapshot ABI but per-
// register name tags on the wire allow safe re-order at read time.
//   - WHvX64RegisterXss (IA32_XSS, 0xDA0): supervisor XSAVE state enable
//   - WHvX64RegisterUCet (IA32_U_CET, 0x6A0): user CET config
//   - WHvX64RegisterSCet (IA32_S_CET, 0x6A2): supervisor CET config
//   - WHvX64RegisterSsp  (IA32_PL3_SSP, mirrors at runtime): user-mode
//     shadow stack pointer (current)
//   - WHvX64RegisterPl0Ssp..Pl3Ssp (IA32_PL0..3_SSP, 0x6A4..0x6A7):
//     per-privilege-level shadow stack pointers
//   - WHvX64RegisterInterruptSspTableAddr (IA32_INTERRUPT_SSP_TABLE_ADDR,
//     0x6A8): interrupt SSP table base
const WHV_REGISTER_NAME kSupervisorMsrNames[] = {
    WHvX64RegisterXss,
    WHvX64RegisterUCet,
    WHvX64RegisterSCet,
    WHvX64RegisterSsp,
    WHvX64RegisterPl0Ssp,
    WHvX64RegisterPl1Ssp,
    WHvX64RegisterPl2Ssp,
    WHvX64RegisterPl3Ssp,
    WHvX64RegisterInterruptSspTableAddr,
};

std::size_t kArchRegCount()    noexcept {
    return sizeof(kArchRegNames)    / sizeof(kArchRegNames[0]);
}
std::size_t kTimingRegCount()  noexcept {
    return sizeof(kTimingRegNames)  / sizeof(kTimingRegNames[0]);
}
std::size_t kIntrCtlRegCount() noexcept {
    return sizeof(kIntrCtlRegNames) / sizeof(kIntrCtlRegNames[0]);
}
std::size_t kSupervisorMsrCount() noexcept {
    return sizeof(kSupervisorMsrNames) / sizeof(kSupervisorMsrNames[0]);
}

void CaptureVcpuState(Vcpu& vp,
                      WHV_PARTITION_HANDLE part,
                      std::uint32_t vp_idx,
                      CapturedVcpuState& out) {
    out.arch.assign(kArchRegCount(), WHV_REGISTER_VALUE{});
    out.timing.assign(kTimingRegCount(), WHV_REGISTER_VALUE{});
    out.intr_ctl.assign(kIntrCtlRegCount(), WHV_REGISTER_VALUE{});
    out.intr_ctl_ok.assign(kIntrCtlRegCount(), false);
    out.sup_msr.assign(kSupervisorMsrCount(), WHV_REGISTER_VALUE{});
    out.sup_msr_ok.assign(kSupervisorMsrCount(), false);
    out.xsave.clear();
    out.apic.clear();

    vp.GetRegisters(
        std::span<const WHV_REGISTER_NAME>(kArchRegNames, kArchRegCount()),
        std::span<WHV_REGISTER_VALUE>(out.arch.data(), out.arch.size()));
    vp.GetRegisters(
        std::span<const WHV_REGISTER_NAME>(kTimingRegNames, kTimingRegCount()),
        std::span<WHV_REGISTER_VALUE>(out.timing.data(), out.timing.size()));
    for (std::size_t i = 0; i < kIntrCtlRegCount(); ++i) {
        try {
            out.intr_ctl[i] = vp.GetRegister(kIntrCtlRegNames[i]);
            out.intr_ctl_ok[i] = true;
        } catch (const std::exception& e) {
            std::fprintf(stderr,
                "[vcpu-state] WARN: get intr_ctl[%zu] (name=0x%08x) failed: %s\n",
                i, static_cast<unsigned>(kIntrCtlRegNames[i]), e.what());
            out.intr_ctl_ok[i] = false;
        }
    }
    // Supervisor MSRs: per-register get with ok bit. WHP rejects names it
    // doesn't implement (older WHP builds without CET support) with
    // HRESULT 0xC0350002 (WHV_E_UNKNOWN_PROPERTY); treat any failure as
    // "register not supported, skip on restore" and continue.
    for (std::size_t i = 0; i < kSupervisorMsrCount(); ++i) {
        try {
            out.sup_msr[i] = vp.GetRegister(kSupervisorMsrNames[i]);
            out.sup_msr_ok[i] = true;
        } catch (const std::exception&) {
            // Silent — these are optional; logging at every probe is noisy.
            out.sup_msr_ok[i] = false;
        }
    }

    UINT32 needed = 0;
    HRESULT hr = WHvGetVirtualProcessorState(
        part, vp_idx, WHvVirtualProcessorStateTypeXsaveState,
        nullptr, 0, &needed);
    if (needed == 0) {
        tinyvmm::Fatal(
            "[vcpu-state] WHvGetVirtualProcessorState(Xsave) query returned 0 bytes");
    }
    (void)hr;
    out.xsave.resize(needed);
    UINT32 actual = 0;
    hr = WHvGetVirtualProcessorState(
        part, vp_idx, WHvVirtualProcessorStateTypeXsaveState,
        out.xsave.data(), needed, &actual);
    tinyvmm::ThrowIfFailed(hr, "WHvGetVirtualProcessorState(Xsave) read");
    out.xsave.resize(actual);

    needed = 0;
    hr = WHvGetVirtualProcessorState(
        part, vp_idx, WHvVirtualProcessorStateTypeInterruptControllerState2,
        nullptr, 0, &needed);
    (void)hr;
    if (needed == 0) {
        out.apic.clear();
        return;
    }
    out.apic.resize(needed);
    hr = WHvGetVirtualProcessorState(
        part, vp_idx, WHvVirtualProcessorStateTypeInterruptControllerState2,
        out.apic.data(), needed, &actual);
    if (FAILED(hr)) {
        std::fprintf(stderr,
            "[vcpu-state] WARN: APIC state read failed (HRESULT=0x%08lX);"
            " skipping APIC capture\n",
            static_cast<unsigned long>(hr));
        out.apic.clear();
        return;
    }
    out.apic.resize(actual);
}

void ApplyVcpuStateNonTiming(Vcpu& vp,
                             WHV_PARTITION_HANDLE part,
                             std::uint32_t vp_idx,
                             const CapturedVcpuState& in) {
    if (in.arch.size() != kArchRegCount() ||
        in.intr_ctl.size() != kIntrCtlRegCount() ||
        in.intr_ctl_ok.size() != kIntrCtlRegCount()) {
        throw std::runtime_error(
            "ApplyVcpuStateNonTiming: CapturedVcpuState vector sizes do not match canonical");
    }
    // M33.7: sup_msr / sup_msr_ok are OPTIONAL. Pre-M33.7 snapshots and
    // pre-M33.7 callers that don't populate the vectors are tolerated:
    // we treat "vectors empty" as "no CET state to restore". When the
    // vectors are present they must match the canonical length so the
    // loop below indexes safely.
    if (in.sup_msr.size() != in.sup_msr_ok.size()) {
        throw std::runtime_error(
            "ApplyVcpuStateNonTiming: sup_msr/sup_msr_ok size mismatch");
    }
    if (!in.sup_msr.empty() && in.sup_msr.size() != kSupervisorMsrCount()) {
        throw std::runtime_error(
            "ApplyVcpuStateNonTiming: sup_msr size does not match canonical");
    }

    // 1. Arch first. Establishes XCR0/CR4/EFER/segments/ApicBase before
    //    XSAVE or APIC blob is applied.
    vp.SetRegisters(
        std::span<const WHV_REGISTER_NAME>(kArchRegNames, kArchRegCount()),
        std::span<const WHV_REGISTER_VALUE>(in.arch.data(), in.arch.size()));

    // 1.5 Supervisor MSRs (M33.7). MUST be applied BEFORE the XSAVE blob
    // so XRSTORS's compact-format validation against IA32_XSS / CET MSRs
    // sees a consistent state. Per-register set with WARN-on-failure so
    // an unsupported name on this WHP build doesn't fail the whole
    // restore. Empty `sup_msr_ok` (e.g. old snapshot without this
    // section) simply skips every register.
    for (std::size_t i = 0; i < in.sup_msr_ok.size(); ++i) {
        if (!in.sup_msr_ok[i]) continue;
        try {
            vp.SetRegister(kSupervisorMsrNames[i], in.sup_msr[i]);
        } catch (const std::exception& e) {
            std::fprintf(stderr,
                "[vcpu-state] WARN: set sup_msr[%zu] (name=0x%08x) failed: %s\n",
                i, static_cast<unsigned>(kSupervisorMsrNames[i]), e.what());
        }
    }

    // 2. XSAVE blob (safe because XCR0 + IA32_XSS + CET MSRs are set).
    //
    // Diagnostic env var (M33.7 keeps for future investigation):
    //   TINYVMM_DEBUG_XSAVE=1   -> log size + first 72 bytes of the blob
    //                              plus XSTATE_BV / XCOMP_BV. Useful when
    //                              kernels start reporting "Bad FPU state"
    //                              after restore on newer hardware that
    //                              advertises new XSAVE components.
    {
        char buf[8] = {};
        if (GetEnvironmentVariableA("TINYVMM_DEBUG_XSAVE", buf, sizeof(buf)) > 0 &&
            buf[0] != '0' && buf[0] != '\0') {
            const std::size_t dump = (std::min)(in.xsave.size(), std::size_t{72});
            std::fprintf(stderr,
                "[vcpu-state] vp=%u xsave size=%zu, first %zu bytes:\n  ",
                vp_idx, in.xsave.size(), dump);
            for (std::size_t k = 0; k < dump; ++k) {
                std::fprintf(stderr, "%02x ",
                             static_cast<unsigned>(in.xsave[k]));
                if ((k & 15) == 15) std::fprintf(stderr, "\n  ");
            }
            std::fprintf(stderr, "\n");
            if (in.xsave.size() >= 576) {
                std::uint64_t xstate_bv = 0, xcomp_bv = 0;
                std::memcpy(&xstate_bv, in.xsave.data() + 512, 8);
                std::memcpy(&xcomp_bv,  in.xsave.data() + 520, 8);
                std::fprintf(stderr,
                    "[vcpu-state] vp=%u XSTATE_BV=0x%016llx XCOMP_BV=0x%016llx\n",
                    vp_idx,
                    static_cast<unsigned long long>(xstate_bv),
                    static_cast<unsigned long long>(xcomp_bv));
            }
        }
    }
    {
        HRESULT hr_x = WHvSetVirtualProcessorState(
            part, vp_idx, WHvVirtualProcessorStateTypeXsaveState,
            in.xsave.data(),
            tinyvmm::util::checked_int_cast<UINT32>(in.xsave.size()));
        tinyvmm::ThrowIfFailed(hr_x, "WHvSetVirtualProcessorState(Xsave)");
    }

    // 3. APIC blob (safe because ApicBase is set; skipped if empty).
    if (!in.apic.empty()) {
        HRESULT hr = WHvSetVirtualProcessorState(
            part, vp_idx, WHvVirtualProcessorStateTypeInterruptControllerState2,
            in.apic.data(),
            tinyvmm::util::checked_int_cast<UINT32>(in.apic.size()));
        tinyvmm::ThrowIfFailed(hr, "WHvSetVirtualProcessorState(APIC)");
    }

    // 4. Interrupt-control synthetic registers, individually.
    for (std::size_t i = 0; i < kIntrCtlRegCount(); ++i) {
        if (!in.intr_ctl_ok[i]) continue;
        try {
            vp.SetRegister(kIntrCtlRegNames[i], in.intr_ctl[i]);
        } catch (const std::exception& e) {
            std::fprintf(stderr,
                "[vcpu-state] WARN: set intr_ctl[%zu] (name=0x%08x) failed: %s\n",
                i, static_cast<unsigned>(kIntrCtlRegNames[i]), e.what());
        }
    }
}

void ApplyVcpuStateTiming(Vcpu& vp,
                          WHV_PARTITION_HANDLE /*part*/,
                          std::uint32_t /*vp_idx*/,
                          const CapturedVcpuState& in) {
    if (in.timing.size() != kTimingRegCount()) {
        throw std::runtime_error(
            "ApplyVcpuStateTiming: CapturedVcpuState.timing size does not match canonical");
    }
    vp.SetRegisters(
        std::span<const WHV_REGISTER_NAME>(kTimingRegNames, kTimingRegCount()),
        std::span<const WHV_REGISTER_VALUE>(in.timing.data(), in.timing.size()));
}

void ApplyVcpuState(Vcpu& vp,
                    WHV_PARTITION_HANDLE part,
                    std::uint32_t vp_idx,
                    const CapturedVcpuState& in) {
    ApplyVcpuStateNonTiming(vp, part, vp_idx, in);
    // Timing registers LAST. Phase 33.6 SMP production restore skips
    // this back-compat wrapper and calls the two helpers separately so
    // it can apply timing back-to-back across every vCPU.
    ApplyVcpuStateTiming(vp, part, vp_idx, in);
}

}  // namespace tinyvmm::whp::snapshot

File: src/whp/vcpu_state.rs
//! Per-vCPU state capture/apply for save/restore. Port of src/whp/vcpu_state.cpp.
//!
//! Capture order: arch -> timing -> intr_ctl (per-reg, ok bits) -> sup_msr
//! (per-reg, ok bits) -> XSAVE blob -> APIC blob (may be empty).
//! Apply order  : arch -> sup_msr (before XSAVE: XRSTORS validates IA32_XSS/CET)
//! -> XSAVE -> APIC (if non-empty) -> intr_ctl (individually). Timing (Tsc/
//! TscAux) is applied LAST (separate call) to minimize cross-vCPU skew.

#![allow(dead_code)]

use crate::error::{Error, Result};
use crate::whp::vcpu::Vcpu;
use windows_sys::Win32::System::Hypervisor::{
    WHvRegisterInternalActivityState, WHvRegisterInterruptState, WHvRegisterPendingInterruption,
    WHvX64RegisterApicBase, WHvX64RegisterCr0, WHvX64RegisterCr2, WHvX64RegisterCr3,
    WHvX64RegisterCr4, WHvX64RegisterCr8, WHvX64RegisterCs, WHvX64RegisterCstar,
    WHvX64RegisterDeliverabilityNotifications, WHvX64RegisterDr0, WHvX64RegisterDr1,
    WHvX64RegisterDr2, WHvX64RegisterDr3, WHvX64RegisterDr6, WHvX64RegisterDr7, WHvX64RegisterDs,
    WHvX64RegisterEfer, WHvX64RegisterEs, WHvX64RegisterFs, WHvX64RegisterGdtr, WHvX64RegisterGs,
    WHvX64RegisterIdtr, WHvX64RegisterInterruptSspTableAddr, WHvX64RegisterKernelGsBase,
    WHvX64RegisterLdtr, WHvX64RegisterLstar, WHvX64RegisterPat, WHvX64RegisterPl0Ssp,
    WHvX64RegisterPl1Ssp, WHvX64RegisterPl2Ssp, WHvX64RegisterPl3Ssp, WHvX64RegisterR10,
    WHvX64RegisterR11, WHvX64RegisterR12, WHvX64RegisterR13, WHvX64RegisterR14, WHvX64RegisterR15,
    WHvX64RegisterR8, WHvX64RegisterR9, WHvX64RegisterRax, WHvX64RegisterRbp, WHvX64RegisterRbx,
    WHvX64RegisterRcx, WHvX64RegisterRdi, WHvX64RegisterRdx, WHvX64RegisterRflags,
    WHvX64RegisterRip, WHvX64RegisterRsi, WHvX64RegisterRsp, WHvX64RegisterSCet,
    WHvX64RegisterSfmask, WHvX64RegisterSs, WHvX64RegisterSsp, WHvX64RegisterStar,
    WHvX64RegisterSysenterCs, WHvX64RegisterSysenterEip, WHvX64RegisterSysenterEsp,
    WHvX64RegisterTr, WHvX64RegisterTsc, WHvX64RegisterTscAux, WHvX64RegisterUCet,
    WHvX64RegisterXCr0, WHvX64RegisterXss, WHV_REGISTER_NAME,
};

pub const ARCH_REGS: &[WHV_REGISTER_NAME] = &[
    WHvX64RegisterRax,
    WHvX64RegisterRcx,
    WHvX64RegisterRdx,
    WHvX64RegisterRbx,
    WHvX64RegisterRsp,
    WHvX64RegisterRbp,
    WHvX64RegisterRsi,
    WHvX64RegisterRdi,
    WHvX64RegisterR8,
    WHvX64RegisterR9,
    WHvX64RegisterR10,
    WHvX64RegisterR11,
    WHvX64RegisterR12,
    WHvX64RegisterR13,
    WHvX64RegisterR14,
    WHvX64RegisterR15,
    WHvX64RegisterRip,
    WHvX64RegisterRflags,
    WHvX64RegisterEs,
    WHvX64RegisterCs,
    WHvX64RegisterSs,
    WHvX64RegisterDs,
    WHvX64RegisterFs,
    WHvX64RegisterGs,
    WHvX64RegisterLdtr,
    WHvX64RegisterTr,
    WHvX64RegisterIdtr,
    WHvX64RegisterGdtr,
    WHvX64RegisterCr0,
    WHvX64RegisterCr2,
    WHvX64RegisterCr3,
    WHvX64RegisterCr4,
    WHvX64RegisterCr8,
    WHvX64RegisterXCr0,
    WHvX64RegisterDr0,
    WHvX64RegisterDr1,
    WHvX64RegisterDr2,
    WHvX64RegisterDr3,
    WHvX64RegisterDr6,
    WHvX64RegisterDr7,
    WHvX64RegisterEfer,
    WHvX64RegisterKernelGsBase,
    WHvX64RegisterApicBase,
    WHvX64RegisterPat,
    WHvX64RegisterSysenterCs,
    WHvX64RegisterSysenterEsp,
    WHvX64RegisterSysenterEip,
    WHvX64RegisterStar,
    WHvX64RegisterLstar,
    WHvX64RegisterCstar,
    WHvX64RegisterSfmask,
];

pub const TIMING_REGS: &[WHV_REGISTER_NAME] = &[WHvX64RegisterTsc, WHvX64RegisterTscAux];

pub const INTR_CTL_REGS: &[WHV_REGISTER_NAME] = &[
    WHvRegisterPendingInterruption,
    WHvRegisterInterruptState,
    WHvX64RegisterDeliverabilityNotifications,
    WHvRegisterInternalActivityState,
];

pub const SUP_MSR_REGS: &[WHV_REGISTER_NAME] = &[
    WHvX64RegisterXss,
    WHvX64RegisterUCet,
    WHvX64RegisterSCet,
    WHvX64RegisterSsp,
    WHvX64RegisterPl0Ssp,
    WHvX64RegisterPl1Ssp,
    WHvX64RegisterPl2Ssp,
    WHvX64RegisterPl3Ssp,
    WHvX64RegisterInterruptSspTableAddr,
];

#[derive(Default)]
pub struct CapturedVcpuState {
    pub arch: Vec<[u8; 16]>,
    pub timing: Vec<[u8; 16]>,
    pub intr_ctl: Vec<[u8; 16]>,
    pub intr_ctl_ok: Vec<bool>,
    pub sup_msr: Vec<[u8; 16]>,
    pub sup_msr_ok: Vec<bool>,
    pub xsave: Vec<u8>,
    pub apic: Vec<u8>,
}

pub fn capture(vp: &Vcpu) -> Result<CapturedVcpuState> {
    let mut out = CapturedVcpuState {
        arch: vp.get_registers_bytes(ARCH_REGS)?,
        timing: vp.get_registers_bytes(TIMING_REGS)?,
        ..Default::default()
    };

    for &name in INTR_CTL_REGS {
        match vp.get_register_bytes(name) {
            Ok(b) => {
                out.intr_ctl.push(b);
                out.intr_ctl_ok.push(true);
            }
            Err(_) => {
                out.intr_ctl.push([0u8; 16]);
                out.intr_ctl_ok.push(false);
            }
        }
    }
    for &name in SUP_MSR_REGS {
        match vp.get_register_bytes(name) {
            Ok(b) => {
                out.sup_msr.push(b);
                out.sup_msr_ok.push(true);
            }
            Err(_) => {
                out.sup_msr.push([0u8; 16]);
                out.sup_msr_ok.push(false);
            }
        }
    }

    out.xsave = vp.get_xsave()?;
    out.apic = vp.get_interrupt_controller();
    Ok(out)
}

pub fn apply_non_timing(vp: &Vcpu, st: &CapturedVcpuState) -> Result<()> {
    if st.arch.len() != ARCH_REGS.len() {
        return Err(Error::msg("vcpu-state: arch reg count mismatch"));
    }
    // 1. Arch first (establishes XCR0/CR4/EFER/segments/ApicBase).
    vp.set_registers_bytes(ARCH_REGS, &st.arch)?;

    // 1.5 Supervisor MSRs BEFORE XSAVE (XRSTORS validates IA32_XSS/CET). Per-reg
    // with WARN-on-failure (unsupported names on older WHP builds are tolerated).
    for (i, &name) in SUP_MSR_REGS.iter().enumerate() {
        if i < st.sup_msr_ok.len() && st.sup_msr_ok[i] {
            if let Err(e) = vp.set_register_bytes(name, &st.sup_msr[i]) {
                eprintln!("[vcpu-state] WARN: set sup_msr[{i}] (0x{name:08x}) failed: {e}");
            }
        }
    }

    // 2. XSAVE blob.
    vp.set_xsave(&st.xsave)?;

    // 3. APIC blob (if present).
    if !st.apic.is_empty() {
        vp.set_interrupt_controller(&st.apic)?;
    }

    // 4. Interrupt-control synthetic registers, individually (WARN on failure).
    for (i, &name) in INTR_CTL_REGS.iter().enumerate() {
        if i < st.intr_ctl_ok.len() && st.intr_ctl_ok[i] {
            if let Err(e) = vp.set_register_bytes(name, &st.intr_ctl[i]) {
                eprintln!("[vcpu-state] WARN: set intr_ctl[{i}] (0x{name:08x}) failed: {e}");
            }
        }
    }
    Ok(())
}

pub fn apply_timing(vp: &Vcpu, st: &CapturedVcpuState) -> Result<()> {
    if st.timing.len() != TIMING_REGS.len() {
        return Err(Error::msg("vcpu-state: timing reg count mismatch"));
    }
    vp.set_registers_bytes(TIMING_REGS, &st.timing)
}

// ============================================================
// Wire encode / decode (name-tagged so reorder-safe at read time)
// ============================================================
fn put_u32(o: &mut Vec<u8>, v: u32) {
    o.extend_from_slice(&v.to_le_bytes());
}

/// [u32 vp_idx][u32 count][ u32 name | 16 bytes value ]*
pub fn encode_regs(vp_idx: u32, names: &[WHV_REGISTER_NAME], vals: &[[u8; 16]]) -> Vec<u8> {
    let mut o = Vec::with_capacity(8 + names.len() * 20);
    put_u32(&mut o, vp_idx);
    put_u32(&mut o, names.len() as u32);
    for (i, &name) in names.iter().enumerate() {
        put_u32(&mut o, name as u32);
        o.extend_from_slice(&vals[i]);
    }
    o
}

/// [u32 vp_idx][u32 count][ u32 name | u8 ok | u8 pad[3] | 16 bytes value ]*
pub fn encode_okregs(
    vp_idx: u32,
    names: &[WHV_REGISTER_NAME],
    vals: &[[u8; 16]],
    ok: &[bool],
) -> Vec<u8> {
    let mut o = Vec::with_capacity(8 + names.len() * 24);
    put_u32(&mut o, vp_idx);
    put_u32(&mut o, names.len() as u32);
    for (i, &name) in names.iter().enumerate() {
        put_u32(&mut o, name as u32);
        o.push(ok[i] as u8);
        o.extend_from_slice(&[0u8, 0, 0]);
        o.extend_from_slice(&vals[i]);
    }
    o
}

/// [u32 vp_idx][u32 len][bytes]
pub fn encode_blob(vp_idx: u32, blob: &[u8]) -> Vec<u8> {
    let mut o = Vec::with_capacity(8 + blob.len());
    put_u32(&mut o, vp_idx);
    put_u32(&mut o, blob.len() as u32);
    o.extend_from_slice(blob);
    o
}

fn rd_u32(b: &[u8], off: usize) -> Result<u32> {
    b.get(off..off + 4)
        .map(|s| u32::from_le_bytes(s.try_into().unwrap()))
        .ok_or_else(|| Error::msg("vcpu-state: truncated section"))
}

/// Decode a reg block into values ordered per `canonical` (default [0;16] for
/// any name absent from the payload). Returns (vp_idx, values).
pub fn decode_regs(
    payload: &[u8],
    canonical: &[WHV_REGISTER_NAME],
) -> Result<(u32, Vec<[u8; 16]>)> {
    let vp_idx = rd_u32(payload, 0)?;
    let count = rd_u32(payload, 4)? as usize;
    let mut map: std::collections::HashMap<u32, [u8; 16]> = std::collections::HashMap::new();
    let mut off = 8;
    for _ in 0..count {
        let name = rd_u32(payload, off)?;
        let val: [u8; 16] = payload
            .get(off + 4..off + 20)
            .ok_or_else(|| Error::msg("vcpu-state: truncated reg"))?
            .try_into()
            .unwrap();
        map.insert(name, val);
        off += 20;
    }
    let out = canonical
        .iter()
        .map(|n| map.get(&(*n as u32)).copied().unwrap_or([0u8; 16]))
        .collect();
    Ok((vp_idx, out))
}

/// Decode an ok-reg block. Returns (vp_idx, values, ok) ordered per `canonical`.
pub fn decode_okregs(
    payload: &[u8],
    canonical: &[WHV_REGISTER_NAME],
) -> Result<(u32, Vec<[u8; 16]>, Vec<bool>)> {
    let vp_idx = rd_u32(payload, 0)?;
    let count = rd_u32(payload, 4)? as usize;
    let mut map: std::collections::HashMap<u32, ([u8; 16], bool)> =
        std::collections::HashMap::new();
    let mut off = 8;
    for _ in 0..count {
        let name = rd_u32(payload, off)?;
        let ok = *payload
            .get(off + 4)
            .ok_or_else(|| Error::msg("vcpu-state: truncated okreg"))?
            != 0;
        let val: [u8; 16] = payload
            .get(off + 8..off + 24)
            .ok_or_else(|| Error::msg("vcpu-state: truncated okreg val"))?
            .try_into()
            .unwrap();
        map.insert(name, (val, ok));
        off += 24;
    }
    let mut vals = Vec::with_capacity(canonical.len());
    let mut oks = Vec::with_capacity(canonical.len());
    for n in canonical {
        match map.get(&(*n as u32)) {
            Some((v, ok)) => {
                vals.push(*v);
                oks.push(*ok);
            }
            None => {
                vals.push([0u8; 16]);
                oks.push(false);
            }
        }
    }
    Ok((vp_idx, vals, oks))
}

/// Decode a blob block. Returns (vp_idx, bytes).
pub fn decode_blob(payload: &[u8]) -> Result<(u32, Vec<u8>)> {
    let vp_idx = rd_u32(payload, 0)?;
    let len = rd_u32(payload, 4)? as usize;
    let bytes = payload
        .get(8..8 + len)
        .ok_or_else(|| Error::msg("vcpu-state: truncated blob"))?
        .to_vec();
    Ok((vp_idx, bytes))
}
